src/common/cockpitjsonprint.c

```diff
@@ -183,7 +188,11 @@ FILE *
 cockpit_json_print_open_memfd (const char *name,
                                int         version)
 {
-  int fd = memfd_create ("cockpit login messages", MFD_ALLOW_SEALING | MFD_CLOEXEC);
+  /* current kernels moan about not specifying exec mode */
+  int fd = memfd_create ("cockpit login messages", MFD_ALLOW_SEALING | MFD_CLOEXEC | MFD_NOEXEC_SEAL);
```

Comment:
I think this sort of fallback arrangement is OK, but I don't really understand why you define the symbol to 0 above if you're going to do this check anyway.

Reply:
I've seen it fail both ways -- e.g. on [aarch64](https://download.copr.fedorainfracloud.org/results/packit/cockpit-project-cockpit-19823/fedora-39-aarch64/06871381-cockpit/builder-live.log.gz) the *build* happened in a Fedora 39 mock, so the constant was defined in the include files. But the running kernel didn't support it yet and hence failed at runtime with `EINVAL`. And of course if you build on old kernel, you won't even have the constant.

Reply:
Fair enough.

I don't think we need to support the case of building on an old kernel and running on a new one (and indeed, you also seem to agree with that since you're unwilling to 'guess' the value of the constant) but in case we don't have the constant set, you're basically making the same call twice here (in case of EINVAL).  That definitely seems useless.

I'd be happy to see some kind of an `#ifdef` at this point.  A fairly evil (but common) idiom is something like:

```c
int fd;
#ifdef MFD_NOEXEC_SEAL
  fd = memfd_create(..., other_stuff | MFS_NOEXEC_SEAL)
  if (fd == -1 && errno == EINVAL)
#endif
    fd = memfd_create(..., other_stuff);
...
```

If you don't like that then i recommend moving the memfd creation to a separate function like:

```C
static int
create_noexec_memfd (void)
{
#ifdef MFD_NOEXEC_SEAL
  int fd = memfd_create(..., other_stuff | MFS_NOEXEC_SEAL)
  if (fd == -1 || errno != EINVAL)
    return fd;
#endif
  return memfd_create(..., other_stuff);
}
```

and call that.

Of the two, I sort of prefer the first.  It's evil but understandable.